### PR TITLE
feat(profiler): add torch profiler in megatron

### PR DESCRIPTION
**Pull request by anHappyDog**
Feature: add torch profiler in megatron and fix flops calculation

### Description
Implement Pytorch Profiler for megatron's forward and backward, use exact token lengths to calculate flops.

### How has this been tested?
Tested 1.5B, 7B in single machine with 8 A800, profile results is ok.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### Usage

See [Pytorch Profiler tutorial](https://docs.pytorch.org/tutorials/intermediate/tensorboard_profiler_tutorial.html) for usage of profile results.

### CODE REVIEW

File: rlinf/workers/actor/megatron_actor_worker.py [hunk 1]
```diff
@@ -455,56 +463,130 @@ def run_forward_backward(self, batch, forward_only=True):
...
-            )
-            if self.enable_dynamic_batch_size:
-                outputs = torch.cat(forward_outputs, dim=0).to(torch.float32)
-                indices = sum(indices, [])
-                assert len(indices) == outputs.size(0), (
-                    f"{len(indices)} vs. {outputs.size()}"
-                )
-                revert_indices = torch.tensor(
-                    get_reverse_idx(indices), dtype=torch.long
+        if self.use_profiler:
```
[the remark this answers is not in the document]
Comment (andylin-hao) [reply rlinf/workers/actor/megatron_actor_worker.py:493]: I think copilot is right. Add explicit `start` and `stop` method for the profiler instead of using contextmanager. I this way you can write
```
if self.use_profiler:
   self.profiler.start_step()
.... # Original training logic 
if self.use_profiler:
   self.profiler.stop_step()
```

File: rlinf/utils/profiler.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        self.current_action = self.schedule(self.step_counter)
+
+        if self.current_action != ProfilerAction.NONE:
+            temp_profiler = profile(
+                **self.profiler_kwargs,
+                on_trace_ready=self.on_trace_ready
+                if self.current_action == ProfilerAction.RECORD
+                else None,
+            )
+            with temp_profiler:
```
Comment (andylin-hao) [inline rlinf/utils/profiler.py]: The torch profiler can be used without the context expression `profiler.start()` and `profiler.stop()`

File: examples/math/config/qwen2.5-1.5b-single-gpu.yaml [hunk 3]
```diff
@@ -227,6 +228,21 @@ actor:
       tensor_model_parallel_size: ${actor.model.tensor_model_parallel_size}
       pipeline_model_parallel_size: ${actor.model.pipeline_model_parallel_size}
       
+    profiler:
+      output_dir: ${runner.output_dir}/${runner.experiment_name}/profiler
+      activities: ["cpu", "cuda"]
```
Comment (FxxxxU) [inline examples/math/config/qwen2.5-1.5b-single-gpu.yaml:235]: These profiles need to be confirmed. It is necessary to determine whether so many information recording options need to be enabled, as enabling too many options will lead to inaccurate performance tests.

File: rlinf/utils/flops.py [hunk 4]
```diff
@@ -13,32 +13,45 @@
...
 
+import torch
 
-def lmhead_flops(hidden_size, vocab_size, batch_size, seq_length):
-    return 2 * batch_size * hidden_size * vocab_size * seq_length
+# Type alias, supports scalar or tensor (only for length parameters)
+TensorOrScalar = Union[int, float, torch.Tensor]
+
+
+def lmhead_flops(hidden_size, vocab_size, seq_length: TensorOrScalar):
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: Add necessary comments and indicate the calculation formulas to facilitate repeated checks by others.

File: rlinf/utils/flops.py [hunk 5]
```diff
@@ -13,32 +13,45 @@
...
+    """Calculate language model head FLOPs, seq_length supports tensor input, batch_size=1"""
+    return 2 * hidden_size * vocab_size * seq_length
 
 
 def qkv_project_flops(
-    hidden_size, num_attn_heads, num_kv_heads, batch_size, seq_length
+    hidden_size, num_attn_heads, num_kv_heads, seq_length: TensorOrScalar
 ):
+    """Calculate QKV projection FLOPs, seq_length supports tensor input, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: The same issue.

File: rlinf/utils/flops.py [hunk 6]
```diff
@@ -13,32 +13,45 @@
...
     hidden_size_kv = hidden_size // (num_attn_heads // num_kv_heads)
-    return 4 * batch_size * seq_length * hidden_size * (hidden_size + hidden_size_kv)
+    return 2 * seq_length * hidden_size * (hidden_size + 2 * hidden_size_kv)
 
 
-def attention_score_flops(hidden_size, batch_size, seq_length):
-    return 4 * batch_size * seq_length * seq_length * hidden_size
+def wo_projection_flops(hidden_size, seq_length: TensorOrScalar):
+    """Calculate output projection FLOPs, seq_length supports tensor input, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: The same issue.

File: rlinf/utils/flops.py [hunk 7]
```diff
@@ -13,32 +13,45 @@
...
+    return 2 * seq_length * hidden_size * hidden_size
 
 
-def mlp_flops(
-    hidden_size, mlp_intermediate_size, batch_size, seq_length, mlp_type: str = "swiglu"
-):
-    return 6 * batch_size * seq_length * hidden_size * mlp_intermediate_size
+def attention_score_flops(hidden_size, seq_length: TensorOrScalar):
+    """Calculate attention score FLOPs, seq_length supports tensor input, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: The same issue.

File: rlinf/utils/flops.py [hunk 8]
```diff
@@ -13,32 +13,45 @@
...
+    return 4 * hidden_size * seq_length * seq_length
 
 
-def rmsnorm_flops(hidden_size, batch_size, seq_length):
-    return 4 * batch_size * seq_length * hidden_size
+def mlp_flops(hidden_size, mlp_intermediate_size, seq_length: TensorOrScalar):
+    """Calculate MLP FLOPs, seq_length supports tensor input, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: The same issue.

File: rlinf/utils/flops.py [hunk 9]
```diff
@@ -13,32 +13,45 @@
...
+    return 6 * seq_length * hidden_size * mlp_intermediate_size
+
+
+def rmsnorm_flops(hidden_size, seq_length: TensorOrScalar):
+    """Calculate RMSNorm FLOPs, seq_length supports tensor input, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: The same issue.

File: rlinf/utils/flops.py [hunk 10]
```diff
@@ -66,118 +79,74 @@ class FLOPSCalculator:
...
-            + lmhead_flops(hidden_size, vocab_size, batch_size, prompt_length)
-        )
+    def flops_inference(self, seq_length: TensorOrScalar):
+        """Inference phase FLOPs calculation, seq_length supports tensor, batch_size=1"""
+        prefill_total_flops = self._calculate_prefill_flops(prompt_length=seq_length)
+
+        return prefill_total_flops
+
+    def _calculate_prefill_flops(self, prompt_length: TensorOrScalar):
+        """Prefill phase FLOPs calculation, prompt_length supports tensor, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: Necessary information comments are missing, and the meaning of variables is unclear.

File: rlinf/utils/flops.py [hunk 11]
```diff
@@ -66,118 +79,74 @@ class FLOPSCalculator:
...
-                    hidden_size, num_attn_heads, num_kv_heads, batch_size, 1
-                )
-                + mlp_flops(hidden_size, mlp_intermediate_size, batch_size, 1)
-                + 2 * rmsnorm_flops(hidden_size, batch_size, 1)
-            )
-            + lmhead_flops(hidden_size, vocab_size, batch_size, 1)
+    def _calculate_decode_flops(
+        self, prompt_length: TensorOrScalar, decode_length: TensorOrScalar
+    ):
+        """Decode phase FLOPs calculation, prompt_length and decode_length support tensor, batch_size=1"""
```
Comment (FxxxxU) [inline rlinf/utils/flops.py]: Necessary information comments are missing, and the meaning of variables is unclear.

File: rlinf/utils/profiler.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        if self.enabled:
+            self.context.__exit__(None, None, None)
+
+
+class PyTorchProfiler:
+    def __init__(
+        self,
+        output_dir: str = "./profiler_output",
+        activities: List[str] = ["cpu", "cuda"],
+        schedule_wait: int = 5,
```
Comment (FxxxxU) [inline rlinf/utils/profiler.py]: Are there only 10 iterations per training session? Is it necessary to consider the possibility of other training quantities? Also, why is repetition needed? It is recommended not to perform repetition and instead extend the warm-up phase.

File: examples/math/config/qwen2.5-1.5b-grpo-megatron-pipeline.yaml [hunk 13]
```diff
@@ -254,6 +255,21 @@ actor:
       tensor_model_parallel_size: ${actor.model.tensor_model_parallel_size}
       pipeline_model_parallel_size: ${actor.model.pipeline_model_parallel_size}
       
+    profiler:
+      output_dir: ${runner.output_dir}/${runner.experiment_name}/profiler
+      activities: ["cpu", "cuda"]
+      record_shapes: True
```
Comment (FxxxxU) [inline examples/math/config/qwen2.5-1.5b-grpo-megatron-pipeline.yaml]: These profiles need to be confirmed. It is necessary to determine whether so many information recording options need to be enabled, as enabling too many options will lead to inaccurate performance tests.

File: examples/math/config/qwen2.5-1.5b-grpo-megatron.yaml [hunk 14]
```diff
@@ -227,6 +228,21 @@ actor:
       tensor_model_parallel_size: ${actor.model.tensor_model_parallel_size}
       pipeline_model_parallel_size: ${actor.model.pipeline_model_parallel_size}
       
+    profiler:
+      output_dir: ${runner.output_dir}/${runner.experiment_name}/profiler
+      activities: ["cpu", "cuda"]
+      record_shapes: True
```
Comment (FxxxxU) [inline examples/math/config/qwen2.5-1.5b-grpo-megatron.yaml]: These profiles need to be confirmed. It is necessary to determine whether so many information recording options need to be enabled, as enabling too many options will lead to inaccurate performance tests.

File: examples/math/config/qwen2.5-32b-grpo-megatron.yaml [hunk 15]
```diff
@@ -228,6 +229,22 @@ actor:
       process_num: 16 # number of processes to use for checkpointing
       tensor_model_parallel_size: ${actor.model.tensor_model_parallel_size}
       pipeline_model_parallel_size: ${actor.model.pipeline_model_parallel_size}
+  
+    profiler:
+      output_dir: ${runner.output_dir}/${runner.experiment_name}/profiler
+      activities: ["cpu", "cuda"]
+      record_shapes: True
```
Comment (FxxxxU) [inline examples/math/config/qwen2.5-32b-grpo-megatron.yaml]: These profiles need to be confirmed. It is necessary to determine whether so many information recording options need to be enabled, as enabling too many options will lead to inaccurate performance tests.

File: examples/math/config/qwen2.5-7b-grpo-megatron.yaml [hunk 16]
```diff
@@ -229,7 +230,23 @@ actor:
       process_num: 16 # number of processes to use for checkpointing
       tensor_model_parallel_size: ${actor.model.tensor_model_parallel_size}
       pipeline_model_parallel_size: ${actor.model.pipeline_model_parallel_size}
-
+    
+    profiler:
+      output_dir: ${runner.output_dir}/${runner.experiment_name}/profiler
+      activities: ["cpu", "cuda"]
```
Comment (FxxxxU) [inline examples/math/config/qwen2.5-7b-grpo-megatron.yaml:234]: These profiles need to be confirmed. It is necessary to determine whether so many information recording options need to be enabled, as enabling too many options will lead to inaccurate performance tests.

File: rlinf/utils/flops.py (hunk 6 above)
Comment (anHappyDog) [reply rlinf/utils/flops.py]: done

File: rlinf/utils/flops.py (hunk 7 above)
Comment (anHappyDog) [reply rlinf/utils/flops.py]: done

File: rlinf/utils/flops.py (hunk 8 above)
Comment (anHappyDog) [reply rlinf/utils/flops.py]: done

File: rlinf/utils/flops.py (hunk 9 above)
Comment (anHappyDog) [reply rlinf/utils/flops.py]: done

File: rlinf/utils/flops.py (hunk 10 above)
Comment (anHappyDog) [reply rlinf/utils/flops.py]: done.
